pagefaultgames/pokerogue
# [Bug] Adjust how counter attacks target to account for uturn/voltswitch and double battles

#1816




## What are the changes?

Fixes counterattack (Counter, Mirror Coat, Metal Burst, Comeuppance) targeting when the attacker switches out with uturn/volt switch/etc.
Also, Metal Burst and Comeuppance have additional target switching for when the attacker faints in a double battle:
"However, if the last Pokémon to hit the user faints before the user moves, the remaining enemy Pokémon will be hit instead." 
This does not apply to Counter and Mirror Coat
https://bulbapedia.bulbagarden.net/wiki/Metal_Burst_(move)
https://bulbapedia.bulbagarden.net/wiki/Comeuppance_(move)

## Why am I doing these changes?

The moves will fail if hit by U-Turn or switch out attacks, when they should still hit the switched-in opponent.
Additionally, Metal Burst and Comeuppance have their own target redirection that hasn't been implemented prior to this PR.



## What did change?


All counterattack moves will now deal the counter damage to the switched in target from U-Turn/Volt Switch/switch out attacks. 
Metal Burst and Comeuppance will now redirect damage to remaining opponents if the original attacker faints before the counterattack goes off.

### Screenshots/Videos



Before (single battle, fails)

https://github.com/pagefaultgames/pokerogue/assets/62030095/6e6798ee-19f0-406a-9c16-5b7c61a750ac

After (single battle, no longer fails)

https://github.com/pagefaultgames/pokerogue/assets/62030095/dcc250d8-742f-49e3-b5d8-1e116f3dbdda

Before redirect in double battles (fails)

https://github.com/pagefaultgames/pokerogue/assets/62030095/cfb6c943-8553-4a43-b1e3-242e025fa73e

Metal Burst/Comeuppance redirect in double battles

https://github.com/pagefaultgames/pokerogue/assets/62030095/15fba225-2747-4464-99ac-0cc87ea4251c


## How to test the changes?




Use `src/overrides.ts` to give the opponent only U-Turn/Volt Switch
Give your starting pokemon the four counter moves: Counter, Mirror Coat, Metal Burst, Comeuppance
Make sure that you are accounting for Counter/Mirror Coat only countering physical/special damage, respectively.
To test the double battle redirection with Metal Burst and Comeuppance, give the opponent a faster pokemon, use metal burst and have your ally faint the attacker before the counter goes off.

## Checklist
